code/datums/chatmessage.dm

```diff
@@ -255,11 +391,23 @@
 	if (originalSpeaker != src && speaker == src)
 		return
 
-	// Display visual above source
+	var/datum/chatmessage/message_to_use
+	var/text_to_use
+
 	if(runechat_flags & EMOTE_MESSAGE)
-		new /datum/chatmessage(raw_message, speaker, src, message_language, list("emote", "italics"))
+		text_to_use = raw_message
+		spans = list("emote", "italics")
 	else
-		new /datum/chatmessage(lang_treat(speaker, message_language, raw_message, spans, null, TRUE), speaker, src, message_language, spans)
+		text_to_use = lang_treat(speaker, message_language, raw_message, spans, null, TRUE)
+		spans = spans ? spans.Copy() : list()
+
+	message_to_use = SSrunechat.messages_by_creation_string["[text_to_use]-[REF(speaker)]-[message_language]-[list2params(spans)]-[CHAT_MESSAGE_LIFESPAN]-[world.time]"]
```

Comment:
if you're going to generate a key here, and inside /datum/chatmessage, why not pass the key in to avoid needing to do it twice?

Reply:
good idea